main.go

```diff
@@ -616,6 +398,388 @@ func main() {
+					metricsv1.WriteMiddleware(authorization.WithAuthorizers(authZ, rbac.Write, "metrics")),
+				),
+			),
+		)
+	})
+
+	// Logs
+	if cfg.logs.enabled {
+		r.Group(func(r chi.Router) {
+			r.Use(authentication.WithTenantID(t.Name, t.ID))
+			r.Use(authentication.WithTenant)
+
+			r.Use(authN)
+			r.Use(authentication.WithTenantHeader(cfg.logs.tenantHeader, t.ID))
+
+			r.Mount("/api/logs/v1/"+t.Name,
+				stripTenantPrefix("/api/logs/v1",
+					logsv1.NewHandler(
+						cfg.logs.readEndpoint,
+						cfg.logs.tailEndpoint,
+						cfg.logs.writeEndpoint,
+						logsv1.Logger(tCfg.logger),
+						logsv1.Registry(tCfg.reg),
+						logsv1.HandlerInstrumenter(tCfg.ins),
+						logsv1.SpanRoutePrefix("/api/logs/v1/{tenant}"),
+						logsv1.ReadMiddleware(authorization.WithAuthorizers(authZ, rbac.Read, "logs")),
+						logsv1.WriteMiddleware(authorization.WithAuthorizers(authZ, rbac.Write, "logs")),
+					),
+				),
+			)
+		})
+	}
+}
+
+func tennatAuthN(t *tenant) (authentication.TenantOIDCConfig, authentication.MTLSConfig, error) {
+	var (
+		oidcConf authentication.TenantOIDCConfig
+		mTLSConf authentication.MTLSConfig
+	)
+
+	switch {
+	case t.OIDC != nil:
+		oidcConf = authentication.TenantOIDCConfig{
+			Tenant: t.Name,
+			OIDCConfig: authentication.OIDCConfig{
+				ClientID:      t.OIDC.ClientID,
+				ClientSecret:  t.OIDC.ClientSecret,
+				GroupClaim:    t.OIDC.GroupClaim,
+				IssuerCA:      t.OIDC.issuerCA,
+				IssuerURL:     t.OIDC.IssuerURL,
+				RedirectURL:   t.OIDC.RedirectURL,
+				UsernameClaim: t.OIDC.UsernameClaim,
+			},
+		}
+
+	case t.MTLS != nil:
+		mTLSConf = authentication.MTLSConfig{
+			Tenant: t.Name,
+			CAs:    t.MTLS.cas,
+		}
+	default:
+		stdlog.Fatalf("tenant %q must specify either an OIDC or an mTLS configuration", t.Name)
+		err := errors.New("tenant must specify either an OIDC or an mTLS configuration : " + t.Name)
+
+		return authentication.TenantOIDCConfig{}, authentication.MTLSConfig{}, err
+	}
+
+	return oidcConf, mTLSConf, nil
+}
+
+func tenantRateLimits(tCfg *tenantsConfig, t *tenant) []ratelimit.Config {
+	rateLimits := []ratelimit.Config{}
+
+	if t.RateLimits != nil {
+		for _, rl := range t.RateLimits {
+			matcher, err := regexp.Compile(rl.Endpoint)
+			if err != nil {
+				level.Warn(tCfg.logger).Log("msg", "failed to compile matcher for rate limiter", "err", err)
+			}
+
+			rateLimits = append(rateLimits, ratelimit.Config{
+				Tenant:  t.Name,
+				Matcher: matcher,
+				Limit:   rl.Limit,
+				Window:  time.Duration(rl.Window),
+			})
+		}
+	}
+
+	return rateLimits
+}
+
+// Create OIDC middleware for a tenant.
+func createOIDCMiddlware(cfg *config, tCfg *tenantsConfig,
+	tenantName string, oidcConf authentication.TenantOIDCConfig) (http.Handler, authentication.Middleware, error) {
+	var (
+		oidcHandler          http.Handler
+		oidcTenantMiddleware authentication.Middleware
+		err                  error
+	)
+
+	for i := 1; i < cfg.tenantRegistrationRetryCount; i++ {
```

Comment:
I don't think this is the right level of abstraction for this variable. IMO, we should be able to retry the entire tenant registration/addition process. So that would mean wrap the entire `newTenant` func in something that can return an error and be retried. Otherwise this seems like it's too restrictive and will quickly need to be refactored.

Reply:
👍 I was not sure as well (hence I left the draft in draft mode). Now I have implemented basic exponential backoff implementation.  thanks for the suggestion!